### PR TITLE
2.5 - Test Concurrent KVM Creation and Appease the Race Checker

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch follows #9898. 

It removes a nasty Goroutine-unsafe pattern used to check `kvm.StartParams`, instead capturing such parameters with `MockContainer` and exposing them to tests.

Various other Goroutine-safety locking is applied to the test infrastructure.

These tests will not fail regular unit tests, but the race-checking CI job will assert correct behaviour.

## QA steps

- Comment out the locking around the manager call to `EnsureCachedImage` and check that running with `-race` fails.
- Restore the locking and check that it passes.

## Documentation changes

None

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Gorouting Safety Checking for Concurrent KVM Container Creation.](https://trello.com/c/G1CdU5iF/1197-gorouting-safety-checking-for-concurrent-kvm-container-creation)

File: container/kvm/mock/mock-kvm.go [hunk 1]
```diff
@@ -55,22 +57,35 @@ func MockFactory() ContainerFactory {
...
+func (mock *MockContainer) Name() string {
 	return mock.name
 }
 
-func (mock *mockContainer) EnsureCachedImage(params kvm.StartParams) error {
+// imageCacheCalls does nothing more than get incremented by calls to the
+// mocked EnsureCachedImage method below.
+// However, it will cause the race checker to fail if such calls are not made
+// in a Goroutine safe manner.
+var imageCacheCalls int
```
Comment (SimonRichardson) [inline container/kvm/mock/mock-kvm.go]: That's ingenious, I'm glad of the comment.

File: container/testing/common.go [hunk 2]
```diff
@@ -30,8 +30,8 @@ func MockMachineConfig(machineId string) (*instancecfg.InstanceConfig, error) {
 	}
 	err = instanceConfig.SetTools(tools.List{
 		&tools.Tools{
-			Version: version.MustParseBinary("2.3.4-quantal-amd64"),
-			URL:     "http://tools.testing.invalid/2.3.4-quantal-amd64.tgz",
+			Version: version.MustParseBinary("2.5.2-bionic-amd64"),
+			URL:     "http://tools.testing.invalid/2.5.2-bionic-amd64.tgz",
```
Comment (SimonRichardson) [inline container/testing/common.go]: Nice...

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**